torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
   }
 
   CondValue emitIsInstance(const Expr& obj, const Expr& classinfo) {
-    // turn (float, (int, tuple)) into a flat list of types and type kind
-    // category checks: tuple_check = true, types = {float, int}
-    struct GatheredTypes {
-      GatheredTypes(ScriptTypeParser parser) : typeParser_(std::move(parser)) {}
-      void gather(const Expr& classinfo) {
-        if (classinfo.kind() == TK_TUPLE_LITERAL) {
-          for (Expr e : TupleLiteral(classinfo).inputs()) {
-            gather(e);
+    // For use with `unifyTypeList`
+    std::stringstream nowhere;
+
+    Value* lhs_val;
+
+    // Get the types on the right-hand side of the `isinstance`
+    // expression, e.g. `isinstance(x, (float, int))` -> `{float, int}.
+    // While doing this, we also want to decompose any `Union` types,
+    // namely `Union`, `Optional`, and `Number`
+    // NB: We can't use `auto` with this lambda because it contains a
+    // recursive call
+    std::function<void(const Expr&, std::vector<TypePtr>&, bool)> gather_types;
+
+    gather_types =
+        [&](const Expr& expr, std::vector<TypePtr>& types, bool is_lhs) {
+          TypePtr t;
+
+          if (!is_lhs) {
+            if (expr.kind() == TK_TUPLE_LITERAL) {
+              for (Expr e : TupleLiteral(expr).inputs()) {
+                gather_types(e, types, is_lhs);
+              }
+              return;
+            }
+            t = typeParser_.parseTypeFromExpr(expr);
+          } else {
+            lhs_val = emitExpr(expr);
+            t = lhs_val->type();
           }
-          return;
-        }
-        TypePtr type = typeParser_.parseTypeFromExpr(classinfo);
-        types.emplace_back(type);
-      }
-      bool staticallyTrue(const TypePtr& actual_type) {
-        // is this isinstance check statically true?
-        for (const TypePtr& typ : types) {
-          if (actual_type->isSubtypeOf(typ)) {
-            return true;
+
+          if (t->kind() == UnionType::Kind) {
+            auto inner = t->expect<UnionType>()->containedTypes().vec();
+            types.insert(types.end(), inner.begin(), inner.end());
+          } else if (t->kind() == OptionalType::Kind) {
+            auto inner = t->expect<OptionalType>()->getElementType();
+            types.push_back(inner);
+            types.push_back(NoneType::get());
+          } else if (t->kind() == NumberType::Kind) {
+            types.push_back(IntType::get());
+            types.push_back(FloatType::get());
+            types.push_back(ComplexType::get());
+          } else {
+            types.push_back(t);
           }
+        };
+
+    auto dedup = [&](std::vector<TypePtr>& types) {
```

Comment:
This clang-tidy warning seems serious. Could you use this `dedup` in code and also add test case to make sure if `dedup` is not called that test fails.